src/cfamily/ondemand.ts

```diff
@@ -0,0 +1,109 @@
+/* --------------------------------------------------------------------------------------------
+ * SonarLint for VisualStudio Code
+ * Copyright (C) 2017-2025 SonarSource SA
+ * sonarlint@sonarsource.com
+ * Licensed under the LGPLv3 License. See LICENSE.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+'use strict';
+
+import * as fs from 'fs';
+import * as openpgp from 'openpgp';
+import * as path from 'path';
+import * as vscode from 'vscode';
+import * as util from '../util/util';
+
+const CFAMILY_PLUGIN_ID = 'sonar-cfamily-plugin';
+const CFAMILY_JAR = 'sonarcfamily.jar';
+
+export function maybeAddCFamilyJar(params: string[]) {
+  const expectedVersion: string = util.packageJson.jarDependencies.filter(dep => dep.artifactId === CFAMILY_PLUGIN_ID)[0].version;
+  const onDemandAnalyzersPath = path.resolve(util.extensionPath, '..', 'sonarsource.sonarlint_ondemand-analyzers');
+  const maybeCFamilyJar = path.resolve(onDemandAnalyzersPath, CFAMILY_PLUGIN_ID, expectedVersion, CFAMILY_JAR);
+  if (fs.existsSync(maybeCFamilyJar)) {
+    params.push(maybeCFamilyJar);
+  } else {
+    // Async call is expected here
+    startDownloadAsync(onDemandAnalyzersPath, expectedVersion);
+  }
+}
+
+async function startDownloadAsync(onDemandAnalyzersPath: string, expectedVersion: string) {
+  let errorMessage = '';
+  const actuallyDownloaded = await vscode.window.withProgress({
+    location: vscode.ProgressLocation.Notification,
+    title: `Downloading ${CFAMILY_PLUGIN_ID} analyzer version ${expectedVersion}`,
+    cancellable: true
+  },  async (progress, cancelToken) => {
+    const fetchAbort = new AbortController();
+    cancelToken.onCancellationRequested(() => {
+      errorMessage = 'Canceled by user';
+      fetchAbort.abort('Download canceled by user');
+    });
+    const url = `https://binaries.sonarsource.com/CommercialDistribution/${CFAMILY_PLUGIN_ID}/${CFAMILY_PLUGIN_ID}-${expectedVersion}.jar`;
+    let fetchResult: Response;
+    try {
+      // On user cancel, AbortController.abort will throw an exception
```

Comment:
What happens if there is a timeout?

Reply:
This should either result in an exception, or in a failed `Response` to `fetch`, so it should be caught below.